src/libraries/System.Threading.RateLimiting/tests/FixedWindowRateLimiterTests.cs

```diff
@@ -0,0 +1,696 @@
+            Assert.Throws<ArgumentOutOfRangeException>(
+                () => new FixedWindowRateLimiterOptions(1, QueueProcessingOrder.NewestFirst, -1, TimeSpan.FromMinutes(2), autoReplenishment: false));
+        }
+
+        [Fact]
+        public override async Task CanAcquireResourceAsync()
+        {
+            var limiter = new FixedWindowRateLimiter(new FixedWindowRateLimiterOptions(1, QueueProcessingOrder.NewestFirst, 1,
+                TimeSpan.Zero, autoReplenishment: false));
+
+            using var lease = await limiter.WaitAsync();
+
+            Assert.True(lease.IsAcquired);
+            var wait = limiter.WaitAsync();
+            Assert.False(wait.IsCompleted);
+
+            Assert.True(limiter.TryReplenish());
+
+            Assert.True((await wait).IsAcquired);
+        }
+
+        [Fact]
+        public override async Task CanAcquireResourceAsync_QueuesAndGrabsOldest()
+        {
+            var limiter = new FixedWindowRateLimiter(new FixedWindowRateLimiterOptions(1, QueueProcessingOrder.OldestFirst, 2,
+                TimeSpan.Zero, autoReplenishment: false));
+            var lease = await limiter.WaitAsync();
+
+            Assert.True(lease.IsAcquired);
+            var wait1 = limiter.WaitAsync();
+            var wait2 = limiter.WaitAsync();
+            Assert.False(wait1.IsCompleted);
+            Assert.False(wait2.IsCompleted);
+
+            lease.Dispose();
+            Assert.True(limiter.TryReplenish());
+
+            lease = await wait1;
+            Assert.True(lease.IsAcquired);
+            Assert.False(wait2.IsCompleted);
+
+            lease.Dispose();
+            Assert.Equal(0, limiter.GetAvailablePermits());
+            Assert.True(limiter.TryReplenish());
+
+            lease = await wait2;
+            Assert.True(lease.IsAcquired);
+        }
+
+        [Fact]
+        public override async Task CanAcquireResourceAsync_QueuesAndGrabsNewest()
+        {
+            var limiter = new FixedWindowRateLimiter(new FixedWindowRateLimiterOptions(2, QueueProcessingOrder.NewestFirst, 3,
+                TimeSpan.FromMinutes(0), autoReplenishment: false));
+
+            var lease = await limiter.WaitAsync(2);
+            Assert.True(lease.IsAcquired);
+
+            var wait1 = limiter.WaitAsync(2);
+            var wait2 = limiter.WaitAsync();
+            Assert.False(wait1.IsCompleted);
+            Assert.False(wait2.IsCompleted);
+
+            lease.Dispose();
+            Assert.True(limiter.TryReplenish());
+
+            // second queued item completes first with NewestFirst
+            lease = await wait2;
+            Assert.True(lease.IsAcquired);
+            Assert.False(wait1.IsCompleted);
+
+            lease.Dispose();
+            Assert.Equal(1, limiter.GetAvailablePermits());
+            Assert.True(limiter.TryReplenish());
+            Assert.True(limiter.TryReplenish());
```

Comment:
nit: remove the second TryReplenish, it makes sense for TokenBucket since it has a replenish rate, but for fixed window it replenishes everything.